lib/ts/BufferWriterFormat.cc

```diff
@@ -336,70 +382,100 @@ namespace bw_fmt
-      if (neg)
-        w.write(neg);
-      if (prefix1) {
-        w.write(prefix1);
-        if (prefix2)
-          w.write(prefix2);
-      }
-      w.write(digits);
-      break;
-    case BWFSpec::Align::CENTER:
-      for (int i = width / 2; i > 0; --i)
-        w.write(spec._fill);
-      if (neg)
-        w.write(neg);
-      if (prefix1) {
-        w.write(prefix1);
-        if (prefix2)
-          w.write(prefix2);
-      }
       w.write(digits);
-      for (int i = (width + 1) / 2; i > 0; --i)
-        w.write(spec._fill);
-      break;
-    case BWFSpec::Align::SIGN:
-      if (neg)
-        w.write(neg);
-      if (prefix1) {
-        w.write(prefix1);
-        if (prefix2)
-          w.write(prefix2);
-      }
-      while (width-- > 0)
-        w.write(spec._fill);
-      w.write(digits);
-      break;
-    default:
-      if (neg)
-        w.write(neg);
-      if (prefix1) {
-        w.write(prefix1);
-        if (prefix2)
-          w.write(prefix2);
+    } else { // use generic Write_Aligned
+      Write_Aligned(w,
+                    [&]() {
+                      if (prefix1) {
+                        w.write(prefix1);
+                        if (prefix2)
+                          w.write(prefix2);
+                      }
+                      w.write(digits);
+                    },
+                    spec._align, width, spec._fill, neg);
+    }
+    return w;
+  }
+
+  /// Format for floating point values. Seperates floating point into a whole number and a
+  /// fraction. The fraction is converted into an unsigned integer based on the specified
+  /// precision, spec._prec. ie. 3.1415 with precision two is seperated into two unsigned
+  /// integers 3 and 14. The different pieces are assembled and placed into the BufferWriter.
+  /// The default is two decimal places. ie. X.XX. The value is always written in base 10.
+  ///
+  /// format: whole.fraction
+  ///     or: left.right
+  BufferWriter &
+  Format_Floating(BufferWriter &w, BWFSpec const &spec, double f, bool neg_p)
+  {
+    uint64_t i = static_cast<uint64_t>(f);
+    if (i == f || spec._prec == 0) { // integral
+      return Format_Integer(w, spec, i, neg_p);
+    }
+
+    static constexpr char dec = '.';
+    double left, right;
+    size_t l = 0;
+    size_t r = 0;
+    char whole[std::numeric_limits<double>::digits10 + 1];
+    char fraction[std::numeric_limits<double>::digits10 + 1];
+    char neg      = 0;
+    int width     = static_cast<int>(spec._min);                             // amount left to fill.
+    int precision = (spec._prec == BWFSpec::DEFAULT._prec) ? 2 : spec._prec; // default precision 2
+
+    right = modf(f, &left); // split the number
+
+    if (neg_p) {
+      neg = '-';
+    } else if (spec._sign != '-') {
+      neg = spec._sign;
+    }
+
+    // Shift the floating point based on the precision. Used to convert
+    //  trailing fraction into an integer value.
+    uint64_t shift = 1;
+    if (precision <= 10) {
```

Comment:
You may want to use `std::array` for the powers table to avoid having a magic constant like `10`. Also, I'd look at something like
```
if (precision < POWERS_OF_TEN.size()) {
  shift = POWERS_OF_TEN[precision];
} else {
  shift = POWERS_OF_TEN.back();
  for ( precision -= POWERS_OF_TEN.size() ; precision > 0 ; --precision )
    shift *= 10;
}
```

Reply:
Updated with `std::array`. Using `POWERS_OF_TEN.size() - 1` to account for 0 precision so that it can index easily.